Datagrid Panel: Edit data within your dashboards

mdvictor opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

Introduces the datagrid panel. A new spreadsheet-like view in which users can edit data within grafana and use it in their panels.

**Why do we need this feature?**

Provides a new way of interacting with a dataset, where users may fine-tune their data before usage. This would prove beneficial to users working a lot with spreadsheets since they will now be able to do their editing within grafana and be more productive.

**Who is this feature for?**

Users that mostly use spreadsheets and spreadsheet-related datasources to pull their data in grafana.

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

mdvictor committed:
wip

mdvictor committed:
Datagrid WIP: snapshotting when data edited, better UI for adding column/rows, refactors

mdvictor committed:
wip

mdvictor committed:
WIP

mdvictor committed:
wip

mdvictor committed:
Add series selector

mdvictor committed:
Delete selection on key press

mdvictor committed:
wip

mdvictor committed:
Multiple row select and delete

mdvictor committed:
wip

mdvictor committed:
draggable rows/columns, better column width calculator

mdvictor committed:
bug fixes

mdvictor committed:
scrollbars

mdvictor committed:
add feature flag

mdvictor committed:
bugfixes

mdvictor committed:
bugfixes

mdvictor committed:
bugfixes

mdvictor committed:
bugfixes

mdvictor committed:
Merge branch 'main' into mdvictor/datagrid

mdvictor committed:
Add possibility to rename column

mdvictor committed:
Input fixes

mdvictor committed:
Merge branch 'main' into mdvictor/datagrid

mdvictor committed:
bugfixes

torkelo commented on public/app/plugins/panel/datagrid/DataGridPanel.tsx:
```diff
@@ -0,0 +1,456 @@
+interface Props extends PanelProps<PanelOptions> {}
+
+export const DataGridPanel: React.FC<Props> = ({ options, data, id, fieldConfig }) => {
+  const [columnWidths, setColumnWidths] = useState<Map<number, number>>(new Map());
+  const [columns, setColumns] = useState<SizedGridColumn[]>([]);
+  const [contextMenuData, setContextMenuData] = useState<DatagridContextMenuData>({ isContextMenuOpen: false });
+  const [renameColumnInputData, setRenameColumnInputData] = useState<RenameColumnInputData>({ isInputOpen: false });
+  const [gridSelection, setGridSelection] = useState<GridSelection>(EMPTY_GRID_SELECTION);
+  const [columnFreezeIndex, setColumnFreezeIndex] = useState<number>(0);
+  const [toggleSearch, setToggleSearch] = useState<boolean>(false);
+  const [isResizeInProgress, setIsResizeInProgress] = useState<boolean>(false);
```
I would try to find a way to move state and logic outside the react componet (using useReducer, or StateManagerBase) or external hook. Easier to unit test state and callback logic and makes it easier to reason about the react rendering logic

mdvictor replied:
I agree with you on state logic. Testing is underway for this, but I wanted to get eyes on it as soon as possible.

baldm0mma commented on public/app/plugins/panel/datagrid/DataGridPanel.tsx:
```diff
@@ -0,0 +1,456 @@
+      [FieldType.other, GridColumnIcon.HeaderReference],
+    ]);
+
+    if (!frame) {
```
I would move this early return to before the `typeToIconMap` declaration, that way, we don't define something that possibly won't be used anyway.

baldm0mma commented on public/app/plugins/panel/datagrid/DataGridPanel.tsx:
```diff
@@ -0,0 +1,456 @@
+
+    const value = field.values.get(row);
+
+    switch (field.type) {
```
Have you considered using an object here instead? iirc switches run linearly, and object key lookups execute in constant time. Plus, I think it'd be easier to test/maintain? Just a thought.

leeoniya replied:
interesting, would be curious to see some benchmarks showing this.

leeoniya replied:
> I'm surprised!

that's the lesson here. any benchmark that's > 6mo old has to be re-evaluated. JITs improve all the time. i've seen a lot :exploding_head: code because some old benchmark showed that the straightforward code was slower 2yrs ago.

leeoniya replied:
actually i think the slowness here is mostly from the lookup version doing function calls, which will always add overhead vs basic constructs like `switch` and `for` (vs forEach)

torkelo commented on public/app/features/query/components/QueryGroup.tsx:
```diff
@@ -94,6 +97,28 @@ export class QueryGroup extends PureComponent<Props, State> {
+    if (config.featureToggles.enableDatagridEditing) {
+      const dashboard = getDashboardSrv().getCurrent();
+      if (dashboard) {
+        this.datagridEditEventSubscription = dashboard.events.getStream(DatagridDataChangeEvent).subscribe({
```
What is this subscription for ?

torkelo replied:
Think we need another way for panels to update query . This feels like the wrong pace / way

mdvictor replied:
No really sure where to set this. The subscription watches for datagrid data changes. Whenever we edit the datagrid panel the snapshot is published to update the dashboard model.

torkelo replied:
@mdvictor yes, but this cannot exist here. something specific to the DataGridPanel cannot live in the global group group component. This needs to be handled from the panel.

torkelo replied:
I think this could be implemented with a new optional method on PanelContext, like onUpdateQueries?: (queries: DataQuery) => void, and then implemented in PanelStateWrapper. 

what do you think @ryantxu @dprokop ?

mdvictor replied:
The main problem is that QueryGroup would not update it's state when the optional method would be called on the PanelContext. I've committed some code to address that. Also, instead of adding a new method on the PanelContext I've found that publishing the `PanelQueriesChangedEvent` event updates the `PanelEditorTabs`.

mdvictor committed:
bugfixes

mdvictor committed:
performance optimisations

baldm0mma commented on public/app/plugins/panel/datagrid/DataGridPanel.tsx:
```diff
@@ -0,0 +1,485 @@
+    return false;
+  };
+
+  const onCellContextMenu = (cell: Item, event: CellClickedEventArgs) => {
```
Seems like onCellContextMenu/onHeaderContextMenu/onHeaderContextMenu has a bunch of shared code, just a nitpick, but maybe refactor into a single function and pass options?

mdvictor replied:
I agree that there's shared code, but creating a single function for this feels like complicating things a bit for just a couple of lines

baldm0mma commented on public/app/plugins/panel/datagrid/components/AddColumn.tsx:
```diff
@@ -0,0 +1,35 @@
+
+import { SimpleInput } from './SimpleInput';
+
+interface Props {
```
I always like naming component props more verbose, in case it ever needs to be exported for testing or referencing in another component to minimize confusion, like `Props ` -> `AddColumnProps `

mdvictor replied:
I agree. fixed

baldm0mma commented on public/app/plugins/panel/datagrid/utils.ts:
```diff
@@ -0,0 +1,183 @@
+};
+
+//Converting an array of nulls or undefineds returns them as strings and prints them in the cells instead of empty cells. Thus the cleanup func
+export const cleanStringFieldAfterConversion = (field: Field): void => {
```
Nice.

baldm0mma commented:
@mdvictor couple of nits, but looks great!

mdvictor committed:
WIP component refactoring and optimisations

mdvictor committed:
comment bit of code to remove error for testing

mdvictor committed:
fix column move and payload types

mdvictor committed:
WIP refactors and tests

torkelo commented on public/app/plugins/panel/datagrid/README.md:
```diff
@@ -0,0 +1,9 @@
+# Dashlist Panel - Native Plugin
+
+This Dashlist panel is **included** with Grafana.
+
+The dashboard list panel allows you to display dynamic links to other dashboards. The list can be configured to use starred dashboards, a search query and/or dashboard tags.
+
+Read more about it here:
+
+[https://grafana.com/docs/grafana/latest/features/panels/dashlist/](https://grafana.com/docs/grafana/latest/features/panels/dashlist/)
```
readme includes copied text, so best be empty?

torkelo commented on public/app/plugins/panel/datagrid/DataGridPanel.tsx:
```diff
@@ -0,0 +1,288 @@
+
+interface DatagridProps extends PanelProps<PanelOptions> {}
+
+export const DataGridPanel: React.FC<DatagridProps> = ({ options, data, id, fieldConfig }) => {
```
```suggestion
export fucntion DataGridPanel({ options, data, id, fieldConfig }: DataGridProips)  {
```

don't use React.FC

torkelo commented on public/app/plugins/panel/datagrid/state.ts:
```diff
@@ -0,0 +1,299 @@
+  [FieldType.other, GridColumnIcon.HeaderReference],
+]);
+
+export const datagridReducer = (state: DatagridState, action: DatagridAction): DatagridState => {
```
This code could be vastly reduced if you used redux toolkit where you can define the reducer and from the reducer actions and action types are automatically inferred .

https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/prometheus/querybuilder/components/PromQueryBuilderContainer.tsx#L66

Oh and you can use redux toolkit with useReducer (as shown in the link above)

torkelo commented on public/app/plugins/panel/datagrid/DataGridPanel.tsx:
```diff
@@ -0,0 +1,288 @@
+  };
+
+  const onRowMove = (from: number, to: number) => {
+    const newFrame = new MutableDataFrame(frame);
+
+    for (const field of newFrame.fields) {
+      const values = field.values.toArray();
+      const value = values[from];
+      values.splice(from, 1);
+      values.splice(to, 0, value);
+      field.values = new ArrayVector(values);
+    }
+
+    publishSnapshot(newFrame, id);
+  };
+
+  const onColumnRename = () => {
+    dispatch({ type: DatagridActionType.showColumnRenameInput });
+  };
+
+  const onRenameInputBlur = (columnName: string, columnIdx: number) => {
+    const newFrame = new MutableDataFrame(frame);
+    newFrame.fields[columnIdx].name = columnName;
+
+    dispatch({ type: DatagridActionType.hideColumnRenameInput });
+    publishSnapshot(newFrame, id);
+  };
+
+  const onSearchClose = () => {
+    dispatch({ type: DatagridActionType.closeSearch });
+  };
+
+  const onGridSelectionChange = (selection: GridSelection) => {
+    dispatch({ type: DatagridActionType.multipleCellsSelected, payload: { selection } });
+  };
+
+  if (!frame) {
+    return <PanelDataErrorView panelId={id} fieldConfig={fieldConfig} data={data} />;
+  }
+
```
I really struggle with these large components with so many callback functions & state logic. 

I think using the StateManagerBase pattern would have worked really nicely here as all these callbacks ALL the state logic could be moved outsider render path (and also tested easily) and be handled together, not this mix of logic and callbacks here and redux actions spreads it out and still leaves a large react component full with mostly inline functions with state logic. etc no need to create callback functions that then dispatch actions, you can bind the react prop directly to the function. 

See example here  https://github.com/grafana/grafana/blob/842ce144292bdf6b51ba2e13961c0986969005e4/public/app/features/dashboard/components/HelpWizard/SupportSnapshotService.ts#L100

But just my preference, I know other see no problem with a render function with a bunch of state callbacks :)

torkelo reviewed (commented):
Beyond the datagrid specific QueryGroup code that think we can better solve with PanelContext function that allows panels to update queries (only really works in this case as they don't have access to the query model so will be a bit of an edge case scenerio). 

But otherwise this looks pretty good (beyond some nitpicks) and the state logic not using redux toolkit (or StateManagerBase which I think would be best for this case with so many callbacks) 

The other thing is that this maybe that the tests for the state logic and the grid itself look a bit thin.

mdvictor committed:
e2e tests

mdvictor committed:
queryGroup subscription refactor

mdvictor committed:
Merge branch 'main' into mdvictor/datagrid

mdvictor commented on public/app/features/query/components/QueryGroup.tsx:
```diff
@@ -130,7 +130,10 @@ export class QueryGroup extends PureComponent<Props, State> {
+      queries: this.props.options.queries,
+      data,
+    });
   }
```
@torkelo I removed the specific Datagrid event from QueryGroup and instead publish a `PanelQueriesChangedEvent`, which updates the PanelEditorTabs. Problem is that when data changes and updates the QueryGroup it doesn't update the queries and thus doesn't update the UI showing that a snapshot is in use. This line fixes that, wdyt?

Using the PanelContext didn't help much because I need to specifically set the state in the queryGroup.

torkelo replied:
@mdvictor I think this could clear the defaults that are applied onmount

torkelo replied:
@mdvictor If I debug this clearly undoes the logic in onmount where query defaults are applied to the initial empty query model
https://github.com/grafana/grafana/blob/main/public/app/features/query/components/QueryGroup.tsx#L103

mdvictor replied:
@torkelo You are right. I've changed the code to update the state properly. wdyt? I added a `componentDidUpdate` so that when the `PanelEditorTabs` get updated by the `PanelQueriesChangedEvent` it also triggers a state update in `QueryGroup`

torkelo replied:
@mdvictor should work, but quite a bit of duplication now between mount and on update code. Would be good to extract that to a shared function

mdvictor committed:
queryGroup - add component on update, fix failing tests

mdvictor committed:
refactor querygroup

torkelo commented on public/app/plugins/panel/datagrid/utils.ts:
```diff
@@ -0,0 +1,304 @@
+
+  panelModel!.refresh();
+
+  dashboard?.events.publish(PanelQueriesChangedEvent);
```
@mdvictor why are you publishing this event here? It's a panel event (published by PanelModel.updateQueries)

mdvictor replied:
That should not be there. will remove

mdvictor committed:
querygroup refactor

mdvictor commented:
Refactored queryGroup and added logic on update. Working on more tests now and then hopefully this is good to go. I'll leave the state logic refactor and datasource change confirmation modal for after the release. They are on the todolist.

torkelo commented on public/app/plugins/panel/datagrid/utils.ts:
```diff
@@ -0,0 +1,301 @@
+    return;
+  }
+
+  const snapshot: DataFrameJSON[] = [dataFrameToJSON(data)];
+  const dashboard = getDashboardSrv().getCurrent();
+  const panelModel = dashboard?.getPanelById(panelID);
+
+  const query: GrafanaQuery = {
+    refId: 'A',
+    queryType: GrafanaQueryType.Snapshot,
+    snapshot,
+    datasource: GRAFANA_DS,
+  };
+
+  panelModel!.updateQueries({
+    dataSource: GRAFANA_DS,
+    queries: [query],
+  });
+
+  panelModel!.refresh();
```
This code is really not allowed, panels can work outside dashboards. getDashboardSrv().getCurrent() can return undefined. This will not work from Scenes apps or new the new dashboard architecture. If the panel needs to communicate / change this outside it needs to be done via PanelContext

torkelo replied:
Panels are rendered in plugins and other pages via PanelRenderer

torkelo replied:
I can try open a PR for a PanelContext extension

mdvictor replied:
I see. Thank you for opening the PR, will have a look over it asap.

mdvictor committed:
tests

mdvictor committed:
fix codeowners validation

mdvictor committed:
lint fixes

mdvictor commented:
We have a pretty good test coverage now and are hopefully close to merging this. All that remains for now imo is deciding how to move forward regarding calling `getDashboardSrv` in the datagrid. I would propose merging the PR by EOD if we don't decide on a path until then and fix it later since the impact is low ( only in scenes, which is in it's early days), unless there are other areas impacted that I've missed. Wdyt @torkelo

torkelo commented:
Nice the tests look great :)

zoltanbedi reviewed (commented):
This looks really good @mdvictor, I haven't found any major thing. I wasn't able to run the e2e tests though.

zoltanbedi commented on public/app/plugins/panel/datagrid/components/RenameColumnCell.tsx:
```diff
@@ -0,0 +1,71 @@
+
+  const onKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
+    if (e.key === 'Enter') {
+      // @ts-ignore
```
+1

mdvictor replied:
Fixed, thank you both!

mdvictor committed:
revert convertFieldType modification in favor of already merged mod + re-add mistakenly deleted line

mdvictor committed:
remove //ts-ignores

mdvictor commented on packages/grafana-data/src/transformations/transformers/convertFieldType.ts:
```diff
@@ -150,7 +150,7 @@ function fieldToNumberField(field: Field): Field {
       // some numbers returned from datasources have commas
       // strip the commas, coerce the string to a number
-      toBeConverted = toBeConverted.replace(/,/g, '');
+      toBeConverted = toBeConverted?.replace(/,/g, '');
```
I reverted the ?. operator and left only the code in main. Thank you for pointing that out!

zoltanbedi replied:
hah, it was fixed by Ryan 😄 I took care of the issues above.

zoltanbedi reviewed (approved):
lgtm, nice work. Let's also wait for @torkelo to approve this as he was involved.

torkelo reviewed (commented):
Just need to figure out the right design for DataGrid should update queries/model 
https://github.com/grafana/grafana/pull/66993

torkelo committed:
Minor style tweaks

torkelo commented:
pushed some minor theme tweaks to align it with the theme in grafana more

mdvictor committed:
fix

mdvictor committed:
align colors with theme

imatwawana reviewed (approved):
Docs update looks good!

mdvictor committed:
Merge branch 'main' into mdvictor/datagrid

mdvictor committed:
fixes

mdvictor commented on packages/grafana-data/src/transformations/transformers/convertFieldType.ts:
```diff
@@ -150,6 +150,10 @@ function fieldToNumberField(field: Field): Field {
+      if (typeof toBeConverted !== 'string') {
+        toBeConverted = String(toBeConverted);
+      }
+
```
There are situations where values are not string in a string field type. e.g: when drag/dropping and csv the dataframe string fields can have both numeric and strings, thus the need for a conversion here. Not sure if fixing it in the convert transformer is the way to go, I think a better approach is to try to convert values when importing, maybe.

mdvictor replied:
refactored this a bit

zoltanbedi replied:
I'm okay with this. Would be nice to have a test for your use case as well.

mdvictor commented on public/app/plugins/panel/datagrid/DataGridPanel.tsx:
```diff
@@ -108,7 +108,11 @@ export function DataGridPanel({ options, data, id, fieldConfig, width, height }:
+
+    newFrame.fields.map((field) => {
+      field.values = new ArrayVector([...field.values.toArray(), null]);
+    });
```
The field values have been modified recently and the received field values aren't ArrayVectors anymore, but ArrayPropertyVector's. I'm not sure why it has changed but the entire datagrid is built around arrayVectors and changing the functionality now is not feasible.

zoltanbedi replied:
Maybe add a TODO message to follow up on this.

mdvictor reviewed (commented):
@torkelo, @zoltanbedi found a couple things I would appreciate your input on.

mdvictor committed:
refactor

mdvictor committed:
Merge branch 'main' into mdvictor/datagrid

mdvictor committed:
add test for convertFieldType transformer and write todo

mdvictor committed:
Merge branch 'main' into mdvictor/datagrid

mdvictor committed:
Merge branch 'main' into mdvictor/datagrid
